app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClient.java

```diff
@@ -239,14 +244,26 @@ static void schedule(Event event) {
         }
 
         /**
-         * If sending is enabled, attempt to send the provided event.
-         *
-         * @param event event
+         * If sending is enabled, attempt to send the provided events.
+         * Also batch the events ordered by their streams, as the QUEUE
+         * can contain events of different streams
+         * @param events list of events
          */
-        static void send(Event event) {
-            postEvent(getStreamConfig(event.getStream()), event);
+        static void send(List<Event> events) {
+            while (!events.isEmpty()) {
+                List<Event> eventsList = new ArrayList<>();
+                String stream = events.get(0).getStream();
+                for (Event event : events) {
+                    if (event.getStream().equals(stream)) {
+                        eventsList.add(event);
+                    }
+                }
+                events.removeAll(eventsList);
```

Comment:
Is it safe to remove elements from `events` while iterating over it with `for-each`?  I would recommend avoiding mutating `events` and instead do something like this (from my [earlier implementation](https://github.com/wikimedia/apps-android-wikipedia/commit/1b9d25f4c1ce3c9a5c8f7da8f5c52ec11b2949cb#diff-806f6b0dfd5d73547d9d1bb0fd0ee4f07cd0e1a5c0f70130e0f1dca5c334df24R238-R261)):

            Map<String, ArrayList<Event>> eventsByStream = new HashMap<>();
            for (Event event : events) {
                String stream = event.getStream();
                if (!eventsByStream.containsKey(stream) || eventsByStream.get(stream) == null) {
                    eventsByStream.put(stream, new ArrayList<>());
                }
                eventsByStream.get(stream).add(event);
            }
            for (String stream : eventsByStream.keySet()) {
                postEvents(getStreamConfig(stream), eventsByStream.get(stream));
            }

Alternatively, if we're not overly concerned with future-proofing at this point, since all events will initially be going to the same event intake service (`eventgate-analytics-external`), we could simplify things by not having `postEvents` check the stream config for the destination URL and instead simply posting the entire queue contents unconditionally to its endpoint URL without any of this preliminary filtering and grouping.

Reply:
I was trying to remove the elements after the for-each was exited @mdholloway . However, segregating them in a map is a better idea, so have changed them. I felt we would miss it in the future if we completely remove this filtering, so implemented per your earlier implementation instead.